### PR TITLE
LEAF-4004 - Inbox v2

### PR DESCRIPTION
**Pull request by mgaoVA**
Prerequisite: https://github.com/department-of-veterans-affairs/LEAF/pull/2124

This PR merge base has been set to PR#2124 to simplify review
- [ ] Update the merge base of this PR to the latest working RC branch after review

This implements a new standard Inbox, replaces the Combined Inbox (which it was mostly copied from), and server-side changes to support efficient loading of new functionality. The old Inbox will remain available from its original link. The key improvement is that the new inbox loads large datasets in smaller chunks, which provides better time-sharing on the server.

By default, the new Inbox organizes actionable results by type of form. An "Organize by Roles" option is available to provide a similar layout to the old Inbox. Admins have a "View as Admin" option to see all requests pending action.

### Summary
- ./api/form/query has a new "join" term: unfilledDependencies. This provides data needed for the Inbox.
- A new option in formQuery.js and intervalQueue.js enables the use of a DOM AbortController to cancel processing
- Form.php -> getIndicator() has a new parameter, forceReadOnly. It defaults to disabled and should only be enabled in safe contexts. This is necessary to keep overall performance equivalent between the new/old inboxes.
- Hot paths have been optimized in Form.php and FormWorkflow.php, reducing database queries.
- Workflow UI headers are more descriptive. For "requestor followup" and "designated group" requirements, the name of the requestor/group is included.

### Potential Impacts
- Dependencies on ./api/form/query
- Dependencies on ./api/formWorkflow/[recordID]/currentStep
- Issues with cache control could cause cached *.js files to prevent the inbox from working

### Testing
- [x] The new Inbox (must click on "View as Admin" and also "Organize by Roles") should have the same number of results as the old Inbox
- [x] Report Builder results as an admin are the same as before
- [x] Report Builder results as a non-admin are the same as before, including data fields within records
- [x] The Workflow section on the "a=printview" page works the same as before

### CODE REVIEW

File: LEAF_Request_Portal/sources/Form.php [hunk 1]
```diff
@@ -1678,13 +1704,21 @@ public function hasWriteAccess($recordID, $categoryID = 0, $indicatorID = 0)
...
-            												AND writable=1', $vars);
+            $resCategoryPrivs = null;
+            $cacheHash = 'hasWriteAccess_catPrivs_'.$categoryID.$this->login->getUserID();
+            if(isset($this->cache[$cacheHash])) {
+                $resCategoryPrivs = $this->cache[$cacheHash];
+            }
+            else {
+                $vars = array(':categoryID' => $categoryID,
+                              ':userID' => $this->login->getUserID());
+                $resCategoryPrivs = $this->db->prepared_query('SELECT * FROM category_privs
```
Comment (Pelentan) [inline LEAF_Request_Portal/sources/Form.php:1715]: Does this need to be a "select *"?

File: LEAF_Request_Portal/sources/Form.php [hunk 2]
```diff
@@ -495,9 +520,8 @@ public function getIndicator($indicatorID, $series, $recordID = null, $parseTemp
                 $form[$idx]['value'] = $this->fileToArray($data[0]['data']);
                 $form[$idx]['raw'] = $data[0]['data'];
             }
-
             // special handling for org chart data types
-            if ($data[0]['format'] == 'orgchart_employee'
+            else if ($data[0]['format'] == 'orgchart_employee'
```
Comment (Pelentan) [inline LEAF_Request_Portal/sources/Form.php:524]: Would like these to be fed through toLower().  We're working on future-proofing things.

File: LEAF_Request_Portal/sources/FormWorkflow.php [hunk 3]
```diff
@@ -218,12 +242,13 @@ public function getActionable(object $form, array $records): array {
...
      * @param object $form instance of Form
      * @param array $records result set from a query on db:records. Requires 'recordID'.
+     * @param bool optional Only select unfilled dependencies
      * @return array $records list of dependencies * records
      */
-    private function retrieveRecordDependencies(object $form, array $records): array {
+    public function getRecordsDependencyData(object $form, array $records, bool $selectUnfilled = false): array {
         $numRecords = count($records);
         if ($numRecords == 0) {
             return $records;
```
Comment (Pelentan) [inline LEAF_Request_Portal/sources/FormWorkflow.php:254]: Would like this method to get modernized a bit more.  There should be only one return.  And we're moving to a json/array format where there are at least two parts.  1 telling the success or fail with data and 1 with the returned data.  How much effort would it be to update this method and its callers?

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

File: LEAF_Request_Portal/templates/reports/LEAF_Inbox.tpl [hunk 4]
```diff
@@ -0,0 +1,926 @@
...
+
+    function buildWorkflowCategoryCache(site) {
+        return $.ajax({
+            type: 'GET',
+            url: site.url + 'api/workflow/categories?x-filterData=categoryID',
+            success: function(res) {
+                res.forEach(w => {
+                    dataWorkflowCategories[w.categoryID] = 1;
+                });
+            }
```
Comment (Pelentan) [inline LEAF_Request_Portal/templates/reports/LEAF_Inbox.tpl:677]: Need a fail/catch

**Review by Pelentan — CHANGES_REQUESTED** [review]

File: LEAF_Request_Portal/sources/Form.php (hunk 1 above)
Comment (mgaoVA) [reply LEAF_Request_Portal/sources/Form.php:1715]: Looks like it'd be safe to change this to a COUNT(*)

File: LEAF_Request_Portal/sources/Form.php (hunk 2 above)
Comment (mgaoVA) [reply LEAF_Request_Portal/sources/Form.php:524]: These should be left as-is. If the format is coming in with a different case, it should be considered wrong.

File: LEAF_Request_Portal/sources/FormWorkflow.php (hunk 3 above)
Comment (mgaoVA) [reply LEAF_Request_Portal/sources/FormWorkflow.php:254]: I think we should formalize a standard response class first, as part of a separate modernization task. Right now the response format is being implemented ad-hoc and can potentially be inconsistent.

File: LEAF_Request_Portal/templates/reports/LEAF_Inbox.tpl (hunk 4 above)
Comment (mgaoVA) [reply LEAF_Request_Portal/templates/reports/LEAF_Inbox.tpl:687]: Updated

**Comment by raphaelfretter** [comment]
Unrelated to actual code review: This would be a good example to use for constructing a PR template in documentation

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by jrodcode — APPROVED** [review]

**Review by Pelentan — APPROVED** [review]

**Review by AmandaSumner — APPROVED** [review]

**Pull request merged**